### DISCUSSION TITLE
UML Sequence Diagram of LXMF Interaction

### DISCUSSION THREAD

**Discussion by LinuxinaBit** [markqvist/Reticulum#408] [Help & Questions] [answered]
Is this sequence diagram at least partially accurate?
I just want to make sure I understand how LXMF works.

The rest of the message has been updated for better accuracy taking into account the replies:
“{}” or “<>” indicates encrypted contents only readable by the receiving party.
“<>” surrounds system messages, “{}” surrounds data messages.
The name before the “:” (i.e. “Alice:”) stands for the sender and their signature.
"[]" symbolizes a Reticulum packet, the “Message for Alice” address is publicly readable by everyone.
That address, the encryption, and the announces are the only things handled by Reticulum, the rest (everything inside of “{}” and “<>”) is handled by LXMF.

![image](https://github.com/markqvist/Reticulum/assets/105567407/b6004cc8-a0d7-486b-ad10-29256a05c8a5)

Sequencediagram.org Source if you want it:
```
note left of Alice: Announces as Peer (TCP)
note left of Bob: Announces as Peer (TCP)
Alice->Bob: [Message for Bob {Alice: Hi Bob}]
note left of Bob: Decrypts and learns sender
Bob->Alice: [Message for Alice <Bob: Recieved>]
note left of Alice: Decrypts and checks signature
note left of Alice: Goes offline
Bob->Alice: [Message for Alice {Bob: Hello Alice}]
note left of Bob: Doesn't get confirmation from Alice
note left of Charlie: Announces as a Replication Node (TCP & LoRa)
Bob->Charlie: [Message for Alice {Bob: Hello Alice}]
note left of David: Announces as Replication Node (LoRa)
Charlie->>David: [Message for Alice {Bob: Hello Alice}]
note left of Alice: Announces as Peer (LoRa)
Alice->>David: [Message for David <Alice: Any new messages?>]
David->>Alice: [Message for Alice {Bob: Hello Alice}]
note left of Alice: Decrypts and learns original sender
note left of Alice: No direct connection to Bob
Alice->>David: [Message for Bob: <Alice: Recieved>]
David->>Charlie: [Message for Bob: <Alice: Recieved>]
Charlie->Bob: [Message for Bob: <Alice: Recieved>]
note left of Bob: Decrypts and checks signature
```

**Comment by LinuxinaBit** [comment]
Incidentally, I have learned that Alice and Bob have lore: https://en.wikipedia.org/wiki/Alice_and_Bob#Cast_of_characters

**Reply by faragher** [reply]
The true question is how much of the lore does Eve know, :D 

For reference, I usually just use Eve as the eavesdropper/attacker and all other names are legitimate users. But communication is collaborative and the only thing that matters is being clear in that instance. A neat read, though.

**Comment by faragher** [comment, marked as answer]
It seems generally correct from my reading. Some general notes:

Reticulum and LXM are different layers. I mention this because Reticulum is always encrypted (and thus so is LXM) but it has no sender information available. This is a feature of links (only if explicitly identified by the sender) and the LXMF format. So with LXM, you are correct, I just want to generalize.

There is no provision for unencrypted messages in Reticulum. Announces are a special case, and local broadcasts are possible, but as a rule, nodes will refuse to route unencrypted traffic.

Alice knows Bob's address but doesn't see his announce, she can't send him anything. She can query the network, and an interface in gateway or AP mode will provide the keys and path if available, and proactively seek it if not.

Sending a message to a propagation node is optional. It depends on the settings of the sender if it should be a fallback, the first choice, or never used at all.

Propagation nodes synchronize messages on a time delay, meaning the only way to ensure timely delivery is to have a common preferred propagation node. Otherwise it's slower than email but faster than snail mail. IIRC, it's a few hours, but I can't recall offhand.

All messages on propagation nodes are 100% encrypted except the destination address. Only the recipient can decrypt any portion of it.

Messages sent over links, which is virtually anything more complicated than a single packet, are encrypted with a symmetric key shared using the asymmetric channel (much like any asymmetric encryption scheme) and as such forward secrecy is assured.

I hope that answered your questions and provided some new insights.

**Reply by LinuxinaBit** [reply]
Ok, that sounds like I thought.
I will update the "key" above the diagram to better explain it though :+1:

**Comment by LinuxinaBit** [comment]
Oh, how does the “On Propagation Net” confirmation work then?

**Reply by faragher** [reply]
That is a very good question I don't have the answer to! 

However, a link can be established without identification, and it's still bidirectional data flow. NomadNet pages, for example, are unidentified by default, but the server can still send data back regarding a request, even asynchronously, until the link is closed.
